README.md
# Laravel Status Codes

A simple Enum to return correct status codes for http responses.

## Installation

Using composer:

```bash
composer require amirhf/laravel-status-code
```

You are then free to use it as needed within your projects.

### Testing

```bash
./vendor/bin/phpunit  
```

## Credits

-   [amirhf](https://github.com/amirhf)

## License

The MIT License (MIT). Please see [License File](LICENSE.md) for more information.

## Laravel Package Boilerplate

This package was generated using the [Laravel Package Boilerplate](https://laravelpackageboilerplate.com).


CHANGELOG.md
# Changelog

All notable changes to `laravel-status-code` will be documented in this file

## 1.0.0 - 201X-XX-XX

- initial release


tests/HttpTest.php
<?php

declare(strict_types=1);

use Amirhf\LaravelStatusCode\Http;
use PHPUnit\Framework\TestCase;

class HttpTest extends TestCase
{
    public function testCorrectStatusCodesAreReturned(): void
    {
        $this->assertEquals(
            100,
            Http::CONTINUE()
        );

        $this->assertEquals(
            101,
            Http::SWITCHING_PROTOCOLS()
        );

        $this->assertEquals(
            102,
            Http::PROCESSING()
        );

        $this->assertEquals(
            103,
            Http::EARLY_HINTS()
        );

        $this->assertEquals(
            200,
            Http::OK()
        );

        $this->assertEquals(
            201,
            Http::CREATED()
        );

        $this->assertEquals(
            202,
            Http::ACCEPTED()
        );

        $this->assertEquals(
            203,
            Http::NON_AUTHORITATIVE_INFORMATION()
        );

        $this->assertEquals(
            204,
            Http::NO_CONTENT()
        );

        $this->assertEquals(
            205,
            Http::RESET_CONTENT()
        );

        $this->assertEquals(
            206,
            Http::PARTIAL_CONTENT()
        );

        $this->assertEquals(
            207,
            Http::MULTI_STATUS()
        );

        $this->assertEquals(
            208,
            Http::ALREADY_REPORTED()
        );

        $this->assertEquals(
            226,
            Http::IM_USED()
        );

        $this->assertEquals(
            300,
            Http::MULTIPLE_CHOICES()
        );

        $this->assertEquals(
            301,
            Http::MOVED_PERMANENTLY()
        );

        $this->assertEquals(
            302,
            Http::FOUND()
        );

        $this->assertEquals(
            303,
            Http::SEE_OTHER()
        );

        $this->assertEquals(
            304,
            Http::NOT_MODIFIED()
        );

        $this->assertEquals(
            305,
            Http::USE_PROXY()
        );

        $this->assertEquals(
            307,
            Http::TEMPORARY_REDIRECT()
        );

        $this->assertEquals(
            308,
            Http::PERMANENT_REDIRECT()
        );

        $this->assertEquals(
            400,
            Http::BAD_REQUEST()
        );

        $this->assertEquals(
            401,
            Http::UNAUTHORIZED()
        );

        $this->assertEquals(
            402,
            Http::PAYMENT_REQUIRED()
        );

        $this->assertEquals(
            403,
            Http::FORBIDDEN()
        );

        $this->assertEquals(
            404,
            Http::NOT_FOUND()
        );

        $this->assertEquals(
            405,
            Http::METHOD_NOT_ALLOWED()
        );

        $this->assertEquals(
            406,
            Http::NOT_ACCEPTABLE()
        );

        $this->assertEquals(
            407,
            Http::PROXY_AUTHENTICATION_REQUIRED()
        );

        $this->assertEquals(
            408,
            Http::REQUEST_TIMEOUT()
        );

        $this->assertEquals(
            409,
            Http::CONFLICT()
        );

        $this->assertEquals(
            410,
            Http::GONE()
        );

        $this->assertEquals(
            411,
            Http::LENGTH_REQUIRED()
        );

        $this->assertEquals(
            412,
            Http::PRECONDITION_FAILED()
        );

        $this->assertEquals(
            413,
            Http::PAYLOAD_TOO_LARGE()
        );

        $this->assertEquals(
            414,
            Http::URI_TOO_LONG()
        );

        $this->assertEquals(
            415,
            Http::UNSUPPORTED_MEDIA_TYPE()
        );

        $this->assertEquals(
            416,
            Http::RANGE_NOT_SATISFIABLE()
        );

        $this->assertEquals(
            417,
            Http::EXPECTATION_FAILED()
        );

        $this->assertEquals(
            418,
            Http::I_AM_A_TEAPOT()
        );

        $this->assertEquals(
            419,
            Http::PAGE_EXPIRED()
        );

        $this->assertEquals(
            421,
            Http::MISDIRECTED_REQUEST()
        );

        $this->assertEquals(
            422,
            Http::UNPROCESSABLE_ENTITY()
        );

        $this->assertEquals(
            423,
            Http::LOCKED()
        );

        $this->assertEquals(
            424,
            Http::FAILED_DEPENDENCY()
        );

        $this->assertEquals(
            425,
            Http::TOO_EARLY()
        );

        $this->assertEquals(
            426,
            Http::UPGRADE_REQUIRED()
        );

        $this->assertEquals(
            428,
            Http::PRECONDITION_REQUIRED()
        );

        $this->assertEquals(
            429,
            Http::TOO_MANY_REQUESTS()
        );

        $this->assertEquals(
            431,
            Http::REQUEST_HEADER_FIELDS_TOO_LARGE()
        );

        $this->assertEquals(
            440,
            Http::LOGIN_TIME_OUT()
        );

        $this->assertEquals(
            444,
            Http::NO_RESPONSE()
        );

        $this->assertEquals(
            449,
            Http::RETRY_WITH()
        );

        $this->assertEquals(
            450,
            Http::BLOCKED_BY_WINDOWS_PARENTAL_CONTROL()
        );

        $this->assertEquals(
            451,
            Http::UNAVAILABLE_FOR_LEGAL_REASONS()
        );

        $this->assertEquals(
            460,
            Http::CLIENT_CLOSED_THE_CONNECTION()
        );

        $this->assertEquals(
            463,
            Http::X_FORWARDED_FOR_TOO_LARGE()
        );

        $this->assertEquals(
            494,
            Http::REQUEST_HEADER_TOO_LARGE()
        );

        $this->assertEquals(
            495,
            Http::SSL_CERTIFICATE_ERROR()
        );

        $this->assertEquals(
            496,
            Http::SSL_CERTIFICATE_REQUIRED()
        );

        $this->assertEquals(
            497,
            Http::HTTP_REQUEST_SENT_TO_HTTPS_PORT()
        );

        $this->assertEquals(
            498,
            Http::INVALID_TOKEN()
        );

        $this->assertEquals(
            499,
            Http::TOKEN_REQUIRED()
        );

        $this->assertEquals(
            500,
            Http::INTERNAL_SERVER_ERROR()
        );

        $this->assertEquals(
            501,
            Http::NOT_IMPLEMENTED()
        );

        $this->assertEquals(
            502,
            Http::BAD_GATEWAY()
        );

        $this->assertEquals(
            503,
            Http::SERVICE_UNAVAILABLE()
        );

        $this->assertEquals(
            504,
            Http::GATEWAY_TIMEOUT()
        );

        $this->assertEquals(
            505,
            Http::HTTP_VERSION_NOT_SUPPORTED()
        );

        $this->assertEquals(
            506,
            Http::VARIANT_ALSO_NEGOTIATES()
        );

        $this->assertEquals(
            507,
            Http::INSUFFICIENT_STORAGE()
        );

        $this->assertEquals(
            508,
            Http::LOOP_DETECTED()
        );

        $this->assertEquals(
            509,
            Http::BANDWIDTH_LIMIT_EXCEEDED()
        );

        $this->assertEquals(
            510,
            Http::NOT_EXTENDED()
        );

        $this->assertEquals(
            511,
            Http::NETWORK_AUTHENTICATION_REQUIRED()
        );

        $this->assertEquals(
            520,
            Http::WEB_SERVER_RETURNED_AN_UNKNOWN_ERROR()
        );

        $this->assertEquals(
            521,
            Http::WEB_SERVER_IS_DOWN()
        );

        $this->assertEquals(
            522,
            Http::CONNECTION_TIMED_OUT()
        );

        $this->assertEquals(
            523,
            Http::ORIGIN_IS_UNREACHABLE()
        );

        $this->assertEquals(
            524,
            Http::A_TIMEOUT_OCCURRED()
        );

        $this->assertEquals(
            525,
            Http::SSL_HANDSHAKE_FAILED()
        );

        $this->assertEquals(
            526,
            Http::INVALID_SSL_CERTIFICATE()
        );

        $this->assertEquals(
            527,
            Http::RAILGUN_ERROR()
        );

        $this->assertEquals(
            529,
            Http::SITE_IS_OVERLOADED()
        );

        $this->assertEquals(
            530,
            Http::SITE_IS_FROZEN()
        );

        $this->assertEquals(
            598,
            Http::NETWORK_READ_TIMEOUT_ERROR()
        );
    }

    public function testErrorIsThrownForInvalidStatusCode(): void
    {
        $this->expectError();
        $this->expectErrorMessage(sprintf('Undefined constant %s::INVALID', Http::class));

        Http::INVALID();
    }
}

src/LaravelStatusCodeServiceProvider.php
<?php

namespace Amirhf\LaravelStatusCode;

use Illuminate\Support\ServiceProvider;

class LaravelStatusCodeServiceProvider extends ServiceProvider
{
    /**
     * Bootstrap the application services.
     */
    public function boot()
    {
        /*
         * Optional methods to load your package assets
         */
        // $this->loadTranslationsFrom(__DIR__.'/../resources/lang', 'laravel-status-code');
        // $this->loadViewsFrom(__DIR__.'/../resources/views', 'laravel-status-code');
        // $this->loadMigrationsFrom(__DIR__.'/../database/migrations');
        // $this->loadRoutesFrom(__DIR__.'/routes.php');

        if ($this->app->runningInConsole()) {
            $this->publishes([
                __DIR__.'/../config/config.php' => config_path('laravel-status-code.php'),
            ], 'config');

            // Publishing the views.
            /*$this->publishes([
                __DIR__.'/../resources/views' => resource_path('views/vendor/laravel-status-code'),
            ], 'views');*/

            // Publishing assets.
            /*$this->publishes([
                __DIR__.'/../resources/assets' => public_path('vendor/laravel-status-code'),
            ], 'assets');*/

            // Publishing the translation files.
            /*$this->publishes([
                __DIR__.'/../resources/lang' => resource_path('lang/vendor/laravel-status-code'),
            ], 'lang');*/

            // Registering package commands.
            // $this->commands([]);
        }
    }

    /**
     * Register the application services.
     */
    public function register()
    {
        // Automatically apply the package configuration
        $this->mergeConfigFrom(__DIR__.'/../config/config.php', 'laravel-status-code');

        // Register the main class to use with the facade
        $this->app->singleton('laravel-status-code', function () {
            return new LaravelStatusCode;
        });
    }
}


src/Http.php
<?php

declare(strict_types=1);

namespace Amirhf\LaravelStatusCode;

use Amirhf\LaravelStatusCode\Exceptions\UndefinedCaseError;

/**
 * @method CONTINUE(): int
 * @method SWITCHING_PROTOCOLS(): int
 * @method PROCESSING(): int
 * @method EARLY_HINTS(): int
 * @method OK(): int
 * @method CREATED(): int
 * @method ACCEPTED(): int
 * @method NON_AUTHORITATIVE_INFORMATION(): int
 * @method NO_CONTENT(): int
 * @method RESET_CONTENT(): int
 * @method PARTIAL_CONTENT(): int
 * @method MULTI_STATUS(): int
 * @method ALREADY_REPORTED(): int
 * @method THIS_IS_FINE(): int
 * @method IM_USED(): int
 * @method MULTIPLE_CHOICES(): int
 * @method MOVED_PERMANENTLY(): int
 * @method FOUND(): int
 * @method SEE_OTHER(): int
 * @method NOT_MODIFIED(): int
 * @method USE_PROXY(): int
 * @method TEMPORARY_REDIRECT(): int
 * @method PERMANENT_REDIRECT(): int
 * @method BAD_REQUEST(): int
 * @method UNAUTHORIZED(): int
 * @method PAYMENT_REQUIRED(): int
 * @method FORBIDDEN(): int
 * @method NOT_FOUND(): int
 * @method METHOD_NOT_ALLOWED(): int
 * @method NOT_ACCEPTABLE(): int
 * @method PROXY_AUTHENTICATION_REQUIRED(): int
 * @method REQUEST_TIMEOUT(): int
 * @method CONFLICT(): int
 * @method GONE(): int
 * @method LENGTH_REQUIRED(): int
 * @method PRECONDITION_FAILED(): int
 * @method PAYLOAD_TOO_LARGE(): int
 * @method URI_TOO_LONG(): int
 * @method UNSUPPORTED_MEDIA_TYPE(): int
 * @method RANGE_NOT_SATISFIABLE(): int
 * @method EXPECTATION_FAILED(): int
 * @method I_AM_A_TEAPOT(): int
 * @method PAGE_EXPIRED(): int
 * @method MISDIRECTED_REQUEST(): int
 * @method UNPROCESSABLE_ENTITY(): int
 * @method LOCKED(): int
 * @method FAILED_DEPENDENCY(): int
 * @method TOO_EARLY(): int
 * @method UPGRADE_REQUIRED(): int
 * @method PRECONDITION_REQUIRED(): int
 * @method TOO_MANY_REQUESTS(): int
 * @method REQUEST_HEADER_FIELDS_TOO_LARGE(): int
 * @method LOGIN_TIME_OUT(): int
 * @method NO_RESPONSE(): int
 * @method RETRY_WITH(): int
 * @method BLOCKED_BY_WINDOWS_PARENTAL_CONTROL(): int
 * @method UNAVAILABLE_FOR_LEGAL_REASONS(): int
 * @method CLIENT_CLOSED_THE_CONNECTION(): int
 * @method X_FORWARDED_FOR_TOO_LARGE(): int
 * @method REQUEST_HEADER_TOO_LARGE(): int
 * @method SSL_CERTIFICATE_ERROR(): int
 * @method SSL_CERTIFICATE_REQUIRED(): int
 * @method HTTP_REQUEST_SENT_TO_HTTPS_PORT(): int
 * @method INVALID_TOKEN(): int
 * @method TOKEN_REQUIRED(): int
 * @method INTERNAL_SERVER_ERROR(): int
 * @method NOT_IMPLEMENTED(): int
 * @method BAD_GATEWAY(): int
 * @method SERVICE_UNAVAILABLE(): int
 * @method GATEWAY_TIMEOUT(): int
 * @method HTTP_VERSION_NOT_SUPPORTED(): int
 * @method VARIANT_ALSO_NEGOTIATES(): int
 * @method INSUFFICIENT_STORAGE(): int
 * @method LOOP_DETECTED(): int
 * @method BANDWIDTH_LIMIT_EXCEEDED(): int
 * @method NOT_EXTENDED(): int
 * @method NETWORK_AUTHENTICATION_REQUIRED(): int
 * @method WEB_SERVER_RETURNED_AN_UNKNOWN_ERROR(): int
 * @method WEB_SERVER_IS_DOWN(): int
 * @method CONNECTION_TIMED_OUT(): int
 * @method ORIGIN_IS_UNREACHABLE(): int
 * @method A_TIMEOUT_OCCURRED(): int
 * @method SSL_HANDSHAKE_FAILED(): int
 * @method INVALID_SSL_CERTIFICATE(): int
 * @method RAILGUN_ERROR(): int
 * @method SITE_IS_OVERLOADED(): int
 * @method SITE_IS_FROZEN(): int
 * @method NETWORK_READ_TIMEOUT_ERROR(): int
 */
enum Http: int
{
    /**
     * @see https://en.wikipedia.org/wiki/List_of_HTTP_status_codes#1xx_informational_response
     */
    case CONTINUE = 100;
    case SWITCHING_PROTOCOLS = 101;
    case PROCESSING = 102;
    case EARLY_HINTS = 103;

    /**
     * @see https://en.wikipedia.org/wiki/List_of_HTTP_status_codes#2xx_success
     */
    case OK = 200;
    case CREATED = 201;
    case ACCEPTED = 202;
    case NON_AUTHORITATIVE_INFORMATION = 203;
    case NO_CONTENT = 204;
    case RESET_CONTENT = 205;
    case PARTIAL_CONTENT = 206;
    case MULTI_STATUS = 207;
    case ALREADY_REPORTED = 208;
    case THIS_IS_FINE = 218;
    case IM_USED = 226;

    /**
     * @see https://en.wikipedia.org/wiki/List_of_HTTP_status_codes#3xx_redirection
     */
    case MULTIPLE_CHOICES = 300;
    case MOVED_PERMANENTLY = 301;
    case FOUND = 302;
    case SEE_OTHER = 303;
    case NOT_MODIFIED = 304;
    case USE_PROXY = 305;
    case TEMPORARY_REDIRECT = 307;
    case PERMANENT_REDIRECT = 308;

    /**
     * @see https://en.wikipedia.org/wiki/List_of_HTTP_status_codes#4xx_client_errors
     */
    case BAD_REQUEST = 400;
    case UNAUTHORIZED = 401;
    case PAYMENT_REQUIRED = 402;
    case FORBIDDEN = 403;
    case NOT_FOUND = 404;
    case METHOD_NOT_ALLOWED = 405;
    case NOT_ACCEPTABLE = 406;
    case PROXY_AUTHENTICATION_REQUIRED = 407;
    case REQUEST_TIMEOUT = 408;
    case CONFLICT = 409;
    case GONE = 410;
    case LENGTH_REQUIRED = 411;
    case PRECONDITION_FAILED = 412;
    case PAYLOAD_TOO_LARGE = 413;
    case URI_TOO_LONG = 414;
    case UNSUPPORTED_MEDIA_TYPE = 415;
    case RANGE_NOT_SATISFIABLE = 416;
    case EXPECTATION_FAILED = 417;
    case I_AM_A_TEAPOT = 418;
    case PAGE_EXPIRED = 419;
    case MISDIRECTED_REQUEST = 421;
    case UNPROCESSABLE_ENTITY = 422;
    case LOCKED = 423;
    case FAILED_DEPENDENCY = 424;
    case TOO_EARLY = 425;
    case UPGRADE_REQUIRED = 426;
    case PRECONDITION_REQUIRED = 428;
    case TOO_MANY_REQUESTS = 429;
    case REQUEST_HEADER_FIELDS_TOO_LARGE = 431;
    case LOGIN_TIME_OUT = 440;
    case NO_RESPONSE = 444;
    case RETRY_WITH = 449;
    case BLOCKED_BY_WINDOWS_PARENTAL_CONTROL = 450;
    case UNAVAILABLE_FOR_LEGAL_REASONS = 451;
    case CLIENT_CLOSED_THE_CONNECTION = 460;
    case X_FORWARDED_FOR_TOO_LARGE = 463;
    case REQUEST_HEADER_TOO_LARGE = 494;
    case SSL_CERTIFICATE_ERROR = 495;
    case SSL_CERTIFICATE_REQUIRED = 496;
    case HTTP_REQUEST_SENT_TO_HTTPS_PORT = 497;
    case INVALID_TOKEN = 498;
    case TOKEN_REQUIRED = 499;

    /**
     * @see https://en.wikipedia.org/wiki/List_of_HTTP_status_codes#5xx_server_errors
     */
    case INTERNAL_SERVER_ERROR = 500;
    case NOT_IMPLEMENTED = 501;
    case BAD_GATEWAY = 502;
    case SERVICE_UNAVAILABLE = 503;
    case GATEWAY_TIMEOUT = 504;
    case HTTP_VERSION_NOT_SUPPORTED = 505;
    case VARIANT_ALSO_NEGOTIATES = 506;
    case INSUFFICIENT_STORAGE = 507;
    case LOOP_DETECTED = 508;
    case BANDWIDTH_LIMIT_EXCEEDED = 509;
    case NOT_EXTENDED = 510;
    case NETWORK_AUTHENTICATION_REQUIRED = 511;
    case WEB_SERVER_RETURNED_AN_UNKNOWN_ERROR = 520;
    case WEB_SERVER_IS_DOWN = 521;
    case CONNECTION_TIMED_OUT = 522;
    case ORIGIN_IS_UNREACHABLE = 523;
    case A_TIMEOUT_OCCURRED = 524;
    case SSL_HANDSHAKE_FAILED = 525;
    case INVALID_SSL_CERTIFICATE = 526;
    case RAILGUN_ERROR = 527;
    case SITE_IS_OVERLOADED = 529;
    case SITE_IS_FROZEN = 530;
    case NETWORK_READ_TIMEOUT_ERROR = 598;

    /**
     * @return int
     */
    public function __invoke(): int
    {
        return $this->value;
    }

    /**
     * @param  string  $name
     * @param  mixed  $args
     * @return int
     *
     * @throws UndefinedCaseError
     */
    public static function __callStatic(string $name, mixed $args): int
    {
        $cases = Http::cases();

        foreach ($cases as $case) {
            if ($case->name === $name) {
                return $case->value;
            }
        }

        throw new UndefinedCaseError(
            enum: Http::class,
            case: $name,
        );
    }

    /**
     * @return array
     */
    public static function options(): array
    {
        return array_column(Http::cases(), 'value', 'name');
    }

    /**
     * @return array
     */
    public static function names(): array
    {
        return array_column(Http::cases(), 'name');
    }

    public static function values(): array
    {
        return array_column(Http::cases(), 'value');
    }
}